pkg/driver/node/mounter/pod_unmounter.go

```diff
@@ -0,0 +1,197 @@
+package mounter
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	crdv1 "github.com/awslabs/aws-s3-csi-driver/pkg/api/v1"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/credentialprovider"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/podmounter/mppod"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/podmounter/mppod/watcher"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/mount-utils"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+)
+
+type PodUnmounter struct {
+	nodeID       string
+	mountUtil    mount.Interface
+	podWatcher   *watcher.Watcher
+	s3paCache    cache.Cache
+	credProvider *credentialprovider.Provider
+}
+
+func NewPodUnmounter(
+	nodeID string,
+	mountUtil mount.Interface,
+	podWatcher *watcher.Watcher,
+	s3paCache cache.Cache,
+	credProvider *credentialprovider.Provider,
+) *PodUnmounter {
+	return &PodUnmounter{
+		nodeID:       nodeID,
+		mountUtil:    mountUtil,
+		podWatcher:   podWatcher,
+		s3paCache:    s3paCache,
+		credProvider: credProvider,
+	}
+}
+
+func (u *PodUnmounter) HandleS3PodAttachmentUpdate(old, new any) {
+	s3pa := new.(*crdv1.MountpointS3PodAttachment)
+	if s3pa.Spec.NodeName != u.nodeID {
+		return
+	}
+
+	for mpPodName, uids := range s3pa.Spec.MountpointS3PodToWorkloadPodUIDs {
+		if len(uids) == 0 {
```

Comment:
Wouldn't this cause race conditions if the node decides to unmount an MP Pod and the controller decides to assign a new workload Pod to this MP Pod? We should ideally mark this MP Pod as deleted from the controller and check that here